libr/anal/labels.c

```diff
@@ -1,98 +1,52 @@
-/* radare - LGPL - Copyright 2014-2018 - pancake */
+/* radare - LGPL - Copyright 2014-2020 - pancake, thestr4ng3r */
 
 #include <r_anal.h>
 
-#define DB anal->sdb_fcns
+HEAPTYPE (ut64);
 
-// list of all the labels for a specific function
-#define LABELS sdb_fmt ("fcn.%"PFMT64x".labels", fcn->addr)
-// value of each element in the labels list
-#define ADDRLABEL(x,y) sdb_fmt ("0x%"PFMT64x"/%s", x,y)
-// resolve by name
-#define LABEL(x) sdb_fmt ("fcn.%"PFMT64x".label.%s", fcn->addr, x)
-// resolve by addr
-#define ADDR(x) sdb_fmt ("fcn.%"PFMT64x".label.0x%"PFMT64x, fcn->addr,x)
-// SDB looks like fcn.0x80480408.labels=0x8048480/patata,0x0405850/potro
-
-R_API ut64 r_anal_fcn_label_get (RAnal *anal, RAnalFunction *fcn, const char *name) {
-	return (anal && fcn)? sdb_num_get (DB, LABEL (name), NULL): UT64_MAX;
+R_API ut64 r_anal_function_get_label(RAnalFunction *fcn, const char *name) {
+	r_return_val_if_fail (fcn, UT64_MAX);
+	ut64 *addr = ht_pp_find (fcn->label_addrs, name, NULL);
+	return addr ? *addr : UT64_MAX;
 }
 
-R_API const char *r_anal_fcn_label_at (RAnal *anal, RAnalFunction *fcn, ut64 addr) {
-	return (anal && fcn)? sdb_const_get (DB, ADDR (addr), NULL): NULL;
+R_API const char *r_anal_function_get_label_at(RAnalFunction *fcn, ut64 addr) {
+	r_return_val_if_fail (fcn, NULL);
+	return ht_up_find (fcn->labels, addr, NULL);
 }
 
-R_API int r_anal_fcn_label_set (RAnal *anal, RAnalFunction *fcn, const char *name, ut64 addr) {
-	if (!anal || !fcn) {
+R_API bool r_anal_function_set_label(RAnalFunction *fcn, const char *name, ut64 addr) {
+	r_return_val_if_fail (fcn && name, false);
+	if (ht_pp_find (fcn->label_addrs, name, NULL)) {
 		return false;
 	}
-	if (sdb_add (DB, ADDR (addr), name, 0)) {
-		if (sdb_num_add (DB, LABEL (name), addr, 0)) {
-			sdb_array_add (DB, LABELS, ADDRLABEL (addr, name), 0);
-			return true;
-		} else {
-			sdb_unset (DB, ADDR (addr), 0);
-		}
-	} else {
-		eprintf ("Cannot add\n");
+	char *n = strdup (name);
+	if (!ht_up_insert (fcn->labels, addr, n)) {
+		free (n);
+		return false;
 	}
-	return false;
+	ht_pp_insert (fcn->label_addrs, name, ut64_new (addr));
```

Comment:
Isn't there a check here because you assume the insert in `label_addrs` cannot fail if the insert in `labels` didn't?

Reply:
The check is here:
https://github.com/radareorg/radare2/pull/17621/files#diff-635a23ca710c8cf67d67f28d75cdcf86R20
```c
if (ht_pp_find (fcn->label_addrs, name, NULL)) {
	return false;
}
```